src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+    move_may_throw() noexcept(false) = default;
+    move_may_throw& operator=(move_may_throw&&) noexcept(false) = default;
+};
+
+// Type trait for determining if move operations on ProtoValue are noexcept.
+// Conditional noexcept-ness of move construction on vector and unordered_map is not
+// guaranteed until c++17, and even then not fully. Thus define a traits class to check whether the
+// containers have nothrow move ops, and use that to determine if ProtoValue's move ops are nothrow.
+// We use move_may_throw as an illustrative dummy class since, perhaps surprisingly, nothrow
+// movability of vector and unordered_map doesn't depend on their stored type.
+struct all_proto_moves_noexcept
+    : std::integral_constant<bool,
+                             std::is_nothrow_move_constructible<std::vector<move_may_throw>>{} &&
+                                 std::is_nothrow_move_constructible<
+                                     std::unordered_map<std::string, move_may_throw>>{}> {};
+
+}  // namespace impl
+
+/// @brief Type-erased value for storing google::protobuf::Value types.
+/// A ProtoValue can be nullptr, bool, int, double, std::string, or, recursively, a vector or
+/// string-map of ProtoValue.
+/// This type is used at API/ABI boundaries for interfacing with grpc/proto code.
+class ProtoValue {
+   public:
+    /// @brief Construct a null object.
+    ProtoValue() noexcept;
+
+    /// @brief Construct a nonempty object.
+    template <typename T>
+    ProtoValue(T t) noexcept(std::is_nothrow_move_constructible<T>{});
+
+    /// @brief Deduction helper constructor for string from string literal
+    ProtoValue(const char* str);
+
+    /// @brief Move construct this from other, leaving other in its unspecified-but-valid moved from
+    /// state.
+    ProtoValue(ProtoValue&& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    ProtoValue(const ProtoValue& other);
+
+    /// @brief Move assignment from other, leaving other in its unspecified-but-valid moved from
+    /// state.
+    ProtoValue& operator=(ProtoValue&& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    ProtoValue& operator=(const ProtoValue& other);
+
+    ~ProtoValue();
+
+    /// @brief Test equality of two types.
+    /// @note "Intuitive" arithmetic equality is not supported, but could be.
+    /// Thus, bool{false}, int{0}, and double{0.0} do not compare equal.
+    friend bool operator==(const ProtoValue& lhs, const ProtoValue& rhs);
+
+    void swap(ProtoValue& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    /// @brief Construct from proto value
+    /// @note This method is trivially templated to insulate google::protobuf::Value from our
+    /// API/ABI. It is meant to be called with no template parameters in a translation unit which
+    /// includes <google/protobuf/struct.pb.h>
+    template <typename Value = google::protobuf::Value>
+    static ProtoValue from_proto(const Value& v);  // NOLINT(misc-no-recursion)
```

Comment:
Do you still need this `NOLINT`?

Reply:
Unfortunately yes https://github.com/viamrobotics/viam-cpp-sdk/actions/runs/10321482772/job/28574485369

I think this may come down to a clang-tidy version thing, running it locally complained about the `NOLINT` in the cpp file at the definition rather than wanting it at the declaration.